### PR TITLE
Atlassian MCP fallback strategy

### PR DESCRIPTION
**Pull request by mrrajan**
- Adds Python REST API client as fallback when Atlassian MCP authentication fails due to organizational policies blocking localhost access. Preserves MCP-first approach with graceful degradation.
- MCP workflows unchanged. REST API only used when MCP fails and user approves.
- Version: 0.5.11 → 0.6.0
- All 5 skills updated:                                                                                                                                                                                            
  - Step 0.5: JIRA Access Initialization with user prompt (Use REST API / Skip JIRA / Retry MCP)                                                                                                                   
  - MCP-first + REST fallback for all JIRA operations
- JIRA TC-4067 and its child issues are created and updated with this fallback strategy.

## Summary by Sourcery

Introduce a JIRA REST API v3 fallback path for all SDLC workflow skills when Atlassian MCP access is unavailable, including a shared Python client, credential handling, and documentation updates.

New Features:
- Add a Python-based JIRA REST API v3 client script used as a fallback when Atlassian MCP operations fail.
- Enable all JIRA-integrated SDLC workflow skills to switch per-operation from MCP to REST API based on user choice and stored credentials.

Enhancements:
- Constrain Bash tool usage in skills to invoking the JIRA REST API client while keeping other operations read-only.
- Extend skill flows with a standardized JIRA access initialization step that handles MCP failure prompts, retries, and REST fallback.
- Document the MCP-first plus REST-fallback strategy, credential storage contract, and token setup across CLAUDE.md and shared docs.

Documentation:
- Describe Atlassian MCP access with REST API v3 fallback in the tools documentation, including credential storage options and supporting guides.
- Update the project configuration contract and CLAUDE.md to define and exemplify the REST API credentials subsection for MCP fallback.
- Add detailed guides for JIRA REST fallback behavior, API token setup, and overall JIRA access strategy for SDLC workflow skills.

### CODE REVIEW

File: .claude-plugin/marketplace.json [hunk 1]
```diff
@@ -11,7 +11,7 @@
     {
       "name": "sdlc-workflow",
       "source": "./plugins/sdlc-workflow",
-      "version": "0.5.11",
+      "version": "0.6.0",
```
Comment (mrizzi) [inline .claude-plugin/marketplace.json]: Please rollback the version. I'll change it later when I'll do the release.

File: CLAUDE.md [hunk 2]
```diff
@@ -36,6 +36,16 @@ When bumping the version, update both files together.
 - Git Pull Request custom field: customfield_10875
 - GitHub Issue custom field: customfield_10747
 
+### REST API Credentials (MCP Fallback)
+
+When Atlassian MCP is unavailable, skills can fall back to JIRA REST API v3. Credentials are optional and only needed if MCP fails.
+
+- Server URL: https://redhat.atlassian.net
+- Email: rravi@redhat.com
```
Comment (mrizzi) [inline CLAUDE.md]: I think it's better to remove them from this file

File: plugins/sdlc-workflow/.claude-plugin/plugin.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "sdlc-workflow",
   "description": "AI-assisted SDLC workflow skills for planning features and implementing tasks with full Jira traceability.",
-  "version": "0.5.11",
+  "version": "0.6.0",
```
Comment (mrizzi) [inline plugins/sdlc-workflow/.claude-plugin/plugin.json]: Same as above.

**Review by mrizzi — COMMENTED** [review]
Some small changes in an overall very good contribution 👏

File: plugins/sdlc-workflow/scripts/jira-client.py [hunk 4]
```diff
@@ -0,0 +1,649 @@
...
+def make_request(method: str, endpoint: str, data: Optional[Dict] = None) -> Dict[str, Any]:
+    """Make authenticated JIRA API request.
+
+    Args:
+        method: HTTP method (GET, POST, PUT, DELETE)
+        endpoint: API endpoint relative to /rest/api/3/
+        data: Optional request body (will be JSON-encoded)
+
+    Returns:
+        Parsed JSON response
```
[the remark this answers is not in the document]
Comment (mrrajan) [reply plugins/sdlc-workflow/scripts/jira-client.py:72]: [sdlc-workflow/verify-pr] Classified as **suggestion** — this proposes adding HTTP timeout configuration as a best practice, but is not documented in CONVENTIONS.md and has no established codebase pattern. No sub-task created.

File: plugins/sdlc-workflow/scripts/jira-client.py [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+def markdown_to_adf(md_text: str) -> Dict[str, Any]:
+    """Convert markdown to Atlassian Document Format.
+
+    Supports:
+    - Headings (# and ##)
+    - Paragraphs
+    - Bullet lists (-)
+    - Ordered lists (1. 2. etc.)
+    - Bold (**text**)
+    - Italic (*text*)
```
[the remark this answers is not in the document]
Comment (mrrajan) [reply plugins/sdlc-workflow/scripts/jira-client.py:157]: [sdlc-workflow/verify-pr] Classified as **code change request** — sub-task [TC-4082](https://redhat.atlassian.net/browse/TC-4082) created to address this feedback.

File: plugins/sdlc-workflow/scripts/jira-client.py [hunk 6]
```diff
@@ -0,0 +1,649 @@
...
+    create_issue_parser.add_argument('--summary', required=True, help='Issue summary')
+    create_issue_parser.add_argument('--description-md', required=True, help='Description in markdown')
+    create_issue_parser.add_argument('--issue-type', required=True, help='Issue type ID or name')
+    create_issue_parser.add_argument('--labels', help='Comma-separated labels')
+    create_issue_parser.add_argument('--assignee-id', help='Assignee account ID')
+
+    # update_issue
+    update_issue_parser = subparsers.add_parser('update_issue', help='Update issue fields')
+    update_issue_parser.add_argument('issue_key', help='Issue key')
+    update_issue_parser.add_argument('--fields-json', required=True, help='Fields as JSON object')
```
[the remark this answers is not in the document]
Comment (mrrajan) [reply plugins/sdlc-workflow/scripts/jira-client.py:616]: [sdlc-workflow/verify-pr] Classified as **code change request** — sub-task [TC-4083](https://redhat.atlassian.net/browse/TC-4083) created to address this feedback.

File: plugins/sdlc-workflow/skills/plan-feature/SKILL.md [hunk 7]
```diff
@@ -75,6 +82,38 @@ If any of these sections are missing or incomplete, inform the user:
 
 **Stop execution immediately.** Do not attempt to gather the missing information or proceed without it.
 
+## Step 0.5 – JIRA Access Initialization
+
+Before attempting any JIRA operations (Steps 1, 4, 6), determine the access method.
+
+**For every JIRA operation:**
+1. **Attempt MCP first** (preferred method)
+2. **If MCP fails, always prompt user:**
```
[the remark this answers is not in the document]
Comment (mrrajan) [reply plugins/sdlc-workflow/skills/plan-feature/SKILL.md:91]: [sdlc-workflow/verify-pr] Classified as **nit** — minor grammar/style feedback that does not affect correctness. No sub-task created.

File: docs/project-config-contract.md [hunk 8]
```diff
@@ -104,6 +125,7 @@ that skills need to create issues, query tasks, and update fields.
 - "Use the Feature issue type ID when creating feature-level issues."
 - "If a Git Pull Request custom field is configured, update it with the PR URL."
 - "If a GitHub Issue custom field is configured, read it from the Jira issue and add a `Closes` reference to the PR description."
+- "If Atlassian MCP fails, always prompt user to use REST API fallback. If user chooses REST API, check for REST API Credentials subsection. If present, use stored credentials; if absent, collect credentials from user and optionally store them."
```
Comment (ruromero) [inline docs/project-config-contract.md]: I think adding the possibility to use an existing `.env` file is better than telling the skill to manage it.
This is what I did from my side for this api key and the Github mcp integration

**Review by ruromero — COMMENTED** [review]
I have only one comment

File: docs/project-config-contract.md (hunk 8 above)
Comment (mrrajan) [reply docs/project-config-contract.md]: Good approach!

File: plugins/sdlc-workflow/scripts/test_jira_client_cli.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    env.pop('JIRA_SERVER_URL', None)
+    env.pop('JIRA_EMAIL', None)
+    env.pop('JIRA_API_TOKEN', None)
+
+    result = subprocess.run(
+        cmd,
+        capture_output=True,
+        text=True,
+        env=env
+    )
```
[the remark this answers is not in the document]
Comment (mrrajan) [reply plugins/sdlc-workflow/scripts/test_jira_client_cli.py]: [sdlc-workflow/verify-pr] Classified as **suggestion** — this is a false positive from static analysis. The flagged code is test infrastructure that calls subprocess.run() with hardcoded test arguments, not user input. The subprocess.run() call uses a list of arguments (not shell=True), which is already safe from shell injection. No sub-task created.

**Comment by mrrajan** [comment]
Thanks @ruromero @mrizzi for your reviews and feedback. I have addressed the changes requested. Please let me know.

**Comment by mrrajan** [comment]
## PR Verification Report

**Task:** TC-4067 — JIRA REST API v3 Fallback for sdlc-workflow  
**PR:** #70  
**Verified by:** sdlc-workflow/verify-pr v0.5.11

### ✅ Review Feedback Resolution

**Status:** PASS

- **Total review threads:** 1 new unclassified thread
- **Classifications:**
  - **Suggestions:** 1 (Sourcery AI security false positive on test code)
  - **Code change requests:** 0
- **Sub-tasks created:** 0
- **Finding:** One new Sourcery AI comment flagged `subprocess.run()` in test infrastructure as potential command injection. Classified as **suggestion** (false positive) because the code is test infrastructure using hardcoded test arguments, not user input. The `subprocess.run()` call uses list arguments (not `shell=True`), which is already safe from shell injection.

### ✅ Scope Containment

**Status:** PASS

All 15 changed files are within the scope of TC-4067:
- Python REST API client (`jira-client.py`) and tests
- Comprehensive documentation (3 new docs in `shared/`)
- Skill updates (5 skills modified to support fallback)
- Project configuration updates (`.env` support, `.gitignore`)
- Contract documentation updates

No out-of-scope changes detected.

### ⚠️ Diff Size

**Status:** WARN (justified)

- **Total diff lines:** 2,836
- **Assessment:** Large diff is expected and justified for a foundational feature that adds:
  - Complete REST API client with stdlib-only dependencies (703 lines)
  - Markdown-to-ADF converter with comprehensive edge case handling
  - 400 lines of tests covering 12 test cases
  - 1,142 lines of implementation documentation (jira-rest-fallback.md, jira-api-token-guide.md, jira-access-strategy.md)
  - Updates to all 5 skill files to integrate fallback logic

### ✅ Commit Traceability

**Status:** PASS

5 commits, all following conventional commit format:
1. `fd837d9` — feat(jira): add REST API v3 fallback when MCP unavailable
2. `27f0062` — update jira client description format
3. `5c9da2d` — fix(jira-client): handle code blocks with blank lines and validate JSON input
4. `4815aa3` — chore: rollback version and remove credentials from CLAUDE.md
5. `c3b6524` — feat: add .env file support for credential management

Each commit is traceable to feature requirements or review feedback.

### ✅ Sensitive Pattern Check

**Status:** PASS

- No hardcoded credentials found
- All token/password references are:
  - Variable names in code
  - Environment variable references (`$JIRA_API_TOKEN`)
  - Documentation explaining secure token handling
- Token masking implemented in `jira-client.py` (`mask_token()` function)

### ⚠️ CI Status

**Status:** WARN (false positive)

- **Sourcery review:** Failed
- **Assessment:** Expected failure — Sourcery AI flagged test infrastructure code as potential security risk. This was manually reviewed and classified as a false positive (see Review Feedback Resolution above).

### ✅ Acceptance Criteria

**Status:** PASS

Feature requirements met:
- ✅ Python REST API client with stdlib-only dependencies
- ✅ Markdown-to-ADF converter built-in
- ✅ All 5 skills updated with fallback logic
- ✅ User prompt flow documented (always prompt on MCP failure)
- ✅ Credential storage with .env support (improvement over original CLAUDE.md-only approach)
- ✅ Token masking in all output
- ✅ HTTP error mapping to user-friendly messages
- ✅ Token rotation flow documented
- ✅ "Skip JIRA" option supported
- ✅ MCP-first approach maintained
- ✅ Comprehensive shared documentation created
- ✅ API token guide copied and adapted
- ✅ Contract documentation updated

**Improvements based on review feedback:**
- Replaced CLAUDE.md credential storage with industry-standard `.env` approach (per ruromero's recommendation)
- Removed credentials from CLAUDE.md (per mrizzi's request)
- Version rollback to 0.5.11 (per mrizzi's request for later release process)

---

**Overall Result:** ✅ PASS

This PR successfully implements the JIRA REST API v3 fallback with comprehensive documentation, test coverage, and security best practices. The large diff is justified by the foundational nature of the feature. All review feedback has been addressed, and one false positive from static analysis has been properly classified.

🤖 _This report was AI-generated by [sdlc-workflow/verify-pr](https://github.com/mrizzi/sdlc-plugins) v0.5.11._

**Review by mrizzi — APPROVED** [review]
@mrrajan thanks a lot 👏
